iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -4,1310 +4,1110 @@
+                        // should never happen. For more details, see the MQTT 3.1.1 specification section "3.2.2.3 Connect Return code"
+                        // https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html
+                        // MQTT 5 supports a larger set of connect codes. See the MQTT 5.0 specification section "3.2.2.2 Connect Reason Code"
+                        // https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901074
+                        throw new IotHubClientException("Failed to open the MQTT connection", cfe);
+                }
             }
         }
 
         public override async Task SendEventAsync(Message message, CancellationToken cancellationToken)
         {
+            cancellationToken.ThrowIfCancellationRequested();
+
             try
             {
-                if (Logging.IsEnabled)
-                    Logging.Enter(this, cancellationToken, nameof(SendEventAsync));
+                // Topic depends on if the client is a module client or a device client
+                string baseTopicName = _moduleId == null ? _deviceToCloudMessagesTopic : _moduleToCloudMessagesTopic;
+                string TopicName = PopulateMessagePropertiesFromMessage(baseTopicName, message);
+
+                if (message.HasPayload && message.Payload.Length > MaxMessageSize)
+                {
+                    throw new InvalidOperationException($"Message size ({message.Payload.Length} bytes) is too big to process. Maximum allowed payload size is {MaxMessageSize}");
+                }
 
-                cancellationToken.ThrowIfCancellationRequested();
+                var mqttMessage = new MqttApplicationMessageBuilder()
+                    .WithTopic(TopicName)
+                    .WithPayload(message.Payload)
+                    .WithQualityOfServiceLevel(publishingQualityOfService)
+                    .Build();
 
-                EnsureValidState();
-                Debug.Assert(_channel != null);
+                MqttClientPublishResult result = await _mqttClient.PublishAsync(mqttMessage, cancellationToken).ConfigureAwait(false);
 
-                await _channel.WriteAndFlushAsync(message).ConfigureAwait(true);
+                if (result.ReasonCode != MqttClientPublishReasonCode.Success)
+                {
+                    throw new IotHubClientException($"Failed to publish the MQTT packet for message with correlation Id {message.CorrelationId} with reason code {result.ReasonCode}", true);
+                }
             }
-            finally
+            catch (Exception ex) when (ex is not IotHubClientException && ex is not InvalidOperationException && ex is not OperationCanceledException)
             {
-                if (Logging.IsEnabled)
-                    Logging.Exit(this, cancellationToken, nameof(SendEventAsync));
+                throw new IotHubClientException("Failed to send message.", true, ex);
```

Comment:
Is there a reason we're deciding to default to retryable? Wouldn't it be better to default to not-retry and retry only for known cases? I say this because of data/bandwidth concerns.
We had a case where the sdk was not returning the correct exception, which should have been non-retryable. I think it was a message-size related error. Instead, the sdk returned the generic communication exception, which was retryable and the customer eventually ran up a huge bill because they were continuously retrying needlessly.

Reply:
Picking a default value here is hard, yeah. I can make this false if we are concerned about cases like that, sure.